### PR TITLE
Variable batch taco2

### PR DESCRIPTION
**Pull request by myagues**
This adds the variable shape batches to Tacotron 2. The return batch is a dictionary instead of a tuple, so it is easier to call specific tensors in the batch and makes order of parameters irrelevant. I have also updated other Tacotron 2 files that use the dataset, let me detail the changes:

- `tacotron_dataset.py`:
  - change tuple batch to dict
  - reads always from raw features and computes normalized features only if `use_norm` is enabled. This means that `stats_path` is a necessary argument when normalizing

- `train_tacotron.py`:
  - simplify `stop_gts`computation with `tf.sequence_mask`, instead of using `max_mel_length`
  - use a common function to reset metrics
  - add Griffin-Lim audio output if `data_config_path`is provided
  - name plots using `utt_id`

- `decode_tacotron2.py`:
  - use a simple dataset for batching, since the class has a lot of boilerplate not necessary for this specific case
  - add option to decode from sentences in a file
  - add audio output

`README` and notebook have also been updated. I've checked training works for both fixed and variable shapes and different values of `reduction factor` [TB.dev](https://tensorboard.dev/experiment/nRZn2R7ERUKVxvMKUvERuw) (the ones without description are variable shape).

I have changed some argument names (`rootdir` -> `data_dir`), because I think are easier to identify, but let me know if I should change it back. Also, some arguments can be added to the config file instead of passing them through command line (`train_dir`, `outdir', etc.)

### CODE REVIEW

**Comment by dathudeptrai** [comment]
@myagues thank you for ur great PR. Let say that this PR will have some confict with the new PR that i will open in this weekend. My pr is a big changes, support multiGPU/singleGPU for base_trainer, all dataset also return batch dictionary rather than tuple. Is that ok if this PR can modify to adapt with my PR ?

**Comment by myagues** [comment]
Yes, no problem, I'll wait and make modifications after you merge your PR.

**Comment by dathudeptrai** [comment]
@myagues seem you are working with Tacotron-2 for a month :)), can you share ur result :v .

**Comment by dathudeptrai** [comment]
@myagues can you solve a confict :v

**Comment by myagues** [comment]
Lots of changes :) Give me some time to solve them.

> seem you have been working with Tacotron-2 for a month :)), can you share ur result :v .

I am not very familiar with TTS other than Tacotron2, so I have been reading the papers for FS and MelGan with the code for the moment.

**Comment by myagues** [comment]
I have updated the changes, everything works on my end.

Please, check that imports work on your end, I think you can get rid of `sys.path.append` by changing the imports.

Mostly cosmetic changes (add GL output, name output files with utt_id), bucket dataset (slightly faster than fixed shape, but may be more noticeable for people with large input sentences) and add text file input in decode.

If you want some further changes, or roll back some others, please let me know.

**Comment by dathudeptrai** [comment]
@myagues thank, i will take a look and merge it :D.

**Comment by myagues** [comment]
I see there are further conflicts, and many changes to the repo since I pulled the request. Let me close this one and split the changes in smaller pulls so it is easier to merge.

**Comment by dathudeptrai** [comment]
@myagues hi, i want to ask you a question. As you see in a dataloader for tacotron2 (https://github.com/TensorSpeech/TensorFlowTTS/blob/master/examples/tacotron2/tacotron_dataset.py#L136-L137). Can we move this np.load to datasets.map(load_fn) so the data loading can be parallel ?.

**Comment by myagues** [comment]
I do not know of any way of loading `npy` files in parallel using such tools. [This SO](https://stackoverflow.com/questions/48889482/feeding-npy-numpy-files-into-tensorflow-data-pipeline) question gives some hints, but we cannot use `tf.data.FixedLengthRecordDataset` since our features are of variable size, and I have not seen any improvement using `tf.pyfunction` wrt sequential read with a `for` loop. Then there is [tfio](https://www.tensorflow.org/io/api_docs/python/tfio/experimental/IODataset#from_numpy_file), but it is designed to be used with a single `npy` file instead of many single ones.

I did a test with `map`, but was unable to use `interleave` because it requires a dataset as the return type.

```python
import os
import time

import numpy as np
import tensorflow as tf


def read_npy_file(item):
    data = np.load(item.numpy().decode())
    return data


file_pattern = os.path.join(
    "/data/ljs_dump/train/norm-feats",
    "*-norm-feats.npy",
)

ds_files = tf.data.Dataset.list_files(file_pattern, seed=0)
ds = ds_files.map(
    lambda item: tf.py_function(read_npy_file, inp=[item], Tout=tf.float32),
    num_parallel_calls=tf.data.experimental.AUTOTUNE
)

t_start = time.perf_counter()
for item in ds_files.take(1000):
    _ = read_npy_file(item)
print("Numpy loop:", time.perf_counter() - t_start)

t_start = time.perf_counter()
for item in ds.take(1000):
    pass
print("TF loop:", time.perf_counter() - t_start)
```

**Comment by dathudeptrai** [comment]
> I do not know of any way of loading `npy` files in parallel using such tools. [This SO](https://stackoverflow.com/questions/48889482/feeding-npy-numpy-files-into-tensorflow-data-pipeline) question gives some hints, but we cannot use `tf.data.FixedLengthRecordDataset` since our features are of variable size, and I have not seen any improvement using `tf.pyfunction` wrt sequential read with a `for` loop. Then there is [tfio](https://www.tensorflow.org/io/api_docs/python/tfio/experimental/IODataset#from_numpy_file), but it is designed to be used with a single `npy` file instead of many single ones.
> 
> I did a test with `map`, but was unable to use `interleave` because it requires a dataset as the return type.
> 
> ```python
> import os
> import time
> 
> import numpy as np
> import tensorflow as tf
> 
> 
> def read_npy_file(item):
>     data = np.load(item.numpy().decode())
>     return data
> 
> 
> file_pattern = os.path.join(
>     "/data/ljs_dump/train/norm-feats",
>     "*-norm-feats.npy",
> )
> 
> ds_files = tf.data.Dataset.list_files(file_pattern, seed=0)
> ds = ds_files.map(
>     lambda item: tf.py_function(read_npy_file, inp=[item], Tout=tf.float32),
>     num_parallel_calls=tf.data.experimental.AUTOTUNE
> )
> 
> t_start = time.perf_counter()
> for item in ds_files.take(1000):
>     _ = read_npy_file(item)
> print("Numpy loop:", time.perf_counter() - t_start)
> 
> t_start = time.perf_counter()
> for item in ds.take(1000):
>     pass
> print("TF loop:", time.perf_counter() - t_start)
> ```

@myagues yeah, i also finish implement it in local, i use tf.numpy_function for `np.load`. I notice that the reading speed is 1x faster :))). So let say, a generator function should just return path, `load and preprocess` function will use dataset.map.
